File: doubly-linked-list.go
package main

// Insertion and deltion can be done in O(1). This means that you can make stacks using a doubly linked lists. It is uncommon but can be done.
// Remember stacks follow LIFO rule
// Linked lists lose to dynamic arrays where random access is required.

// Access: O(n)
// Insertion / Delete at end: O(1)
// Insertion / Delete at middle: O(1) -> this is only if you have access to the element being deleted. Otherwise O(n) will be required to find the element to be deleted.

type DoublylinkedListNode struct {
	Val      int
	Next     *DoublylinkedListNode
	Previous *DoublylinkedListNode
}

func newDoublyNode(val int, prev, next *DoublylinkedListNode) *DoublylinkedListNode {
	return &DoublylinkedListNode{
		Val:      val,
		Next:     next,
		Previous: prev,
	}
}

type DoublyLinkedList struct {
	Head, Tail *DoublylinkedListNode
}

func newDoublyLinkedList() *DoublyLinkedList {
	head := newDoublyNode(-1, nil, nil)  //dummy head node
	tail := newDoublyNode(-1, head, nil) //dummy tail node

	head.Next = tail

	return &DoublyLinkedList{
		Head: head,
		Tail: tail,
	}
}

func (this *DoublyLinkedList) Get(index int) int {
	curr := this.Head.Next

	// keep moving current to left while decrementing the index to find the correct node
	for curr != nil && index > 0 {
		curr = curr.Next
		index--
	}

	if curr != nil && curr != this.Tail && index == 0 {
		return curr.Val
	}

	return -1
}

func (this *DoublyLinkedList) AddAtHead(val int) {
	node := newDoublyNode(val, this.Head, this.Head.Next)
	this.Head.Next.Previous = node
	this.Head.Next = node
}

func (this *DoublyLinkedList) AddAtTail(val int) {
	node := newDoublyNode(val, this.Tail.Previous, this.Tail)
	this.Tail.Previous.Next = node
	this.Tail.Previous = node
}

func (this *DoublyLinkedList) AddAtIndex(val, index int) {
	curr := this.Head.Next

	for curr.Next != nil && index > 0 {
		curr = curr.Next
		index--
	}

	if index == 0 {
		node := newDoublyNode(val, curr.Previous, curr)
		curr.Previous.Next = node
		curr.Previous = node
	}

	return
}

func (this *DoublyLinkedList) DeleteAtIndex(val, index int) {
	curr := this.Head.Next

	for curr.Next != nil && index > 0 {
		curr = curr.Next
		index--
	}

	if index == 0 && curr != nil && curr != this.Tail {
		prev, next := curr.Previous, curr.Next
		prev.Next = next
		next.Previous = prev
	}
	return
}

File: DoublyLinkedList.java
/**
 * Doubly linked list is a variation of the singly linked list.
 * In addition to the next pointer like the singly linked list, the doubly linked list also has a previous pointer.
 * The doubly linked list also has the head and tail just like the singly linked list
 *
 * Inserting at end O(1): The tail would get the new node as its next and the new node will get the current tail as its previous pointer
 * after which the new node will become the new tail.
 *
 * Removing from end O(1): We will get the node previous from the tail using the tails previous pointer and then set the previous node's next to null,
 * after which we will set the node previous from the current tail as the new tail.
 *
 * Access a random element form the linked list O(n): Find the value you are looking for following the pointers
 *
 * Stack can be implemented with linkedLists but it is best to be done with dynamic arrays since random access in dynamic arrays is in O(1) time.
 */
public class DoublyLinkedList {
    ListNode head;
    ListNode tail;

    public static class ListNode {
        int val;
        ListNode next;
        ListNode prev;

        public ListNode(int val) {
            this.val = val;
            this.next = null;
            this.prev = null;
        }
    }

    public DoublyLinkedList() {
        head = new ListNode(-1);
        tail = new ListNode(-1);

        head.next = tail;
        tail.prev = head;
    }

    public void insertFront(int val) {
        ListNode newNode = new ListNode(val);

        // assign correct pointers for the new node
        newNode.prev = head;
        newNode.next = head.next;

        // update prev pointer for the heads next and the heads next pointer
        head.next.prev = newNode;
        head.next = newNode;
    }

    public void insertEnd(int val) {
        ListNode newNode = new ListNode(val);

        // Update new node pointers
        newNode.next = tail;
        newNode.prev = tail.prev;

        // Update tail pointers
        tail.prev.next = newNode;
        tail.prev = newNode;
    }

    public void removeFront() {
        head.next.next.prev = head;
        head.next = head.next.next;
    }

    public void removeEnd() {
        tail.prev.prev.next = tail;
        tail.prev = tail.prev.prev;
    }

    public void printList() {
        ListNode curr = head.next;
        while(curr != null) {
            System.out.println(curr.val + " -> ");
            curr = curr.next;
        }
        System.out.println();
    }
}



